### DISCUSSION TITLE
Help with implementing fluorescent material

### DISCUSSION THREAD

**Discussion by arpit15** [mitsuba-renderer/mitsuba3#913] [Q&A]
I want to implement a basic fluorescent material given a re-radiation matrix. 
**First** idea is to expand the `BSDFContext` struct to add `L_o` or `L_i` based on the transport mode. Then, use this spectrum information to write `eval`, `sample` and `pdf` functions based on re-radiation matrix. However, this requires to change the integrator code to take the processed radiance after interaction with this material instead of multiplying the returned value. 

**Second** idea is to write a custom class which outputs a re-radiation matrix, similar to polarization BSDFs, and write custom multiplication operators for obtain the radiance value after interaction with the material. 

Any help or pointer on how to implement fluorescent material is welcome. 
I am aware of [ART renderer](https://gitlab.com/art-development-team/ART). I will use it to test the accuracy of the implemented material.

**Comment by njroussel** [comment]
Hi @arpit15 

The "cheap" way is to just modify `si.wavelenghts` wherever needed. However, proper importance sampling would require some changes to the interfaces. I haven't really put much thought into it, but your first suggestion seems to be the more natural interface.
